usr.sbin/devinfo/devinfo.c

```diff
@@ -116,52 +197,92 @@ print_device_rman_resources(struct devinfo_rman *rman, void *arg)
 	    print_device_matching_resource, ia) != 0) {
 
 		/* there are, print header */
-		for (i = 0; i < indent; i++)
-			printf(" ");
-		printf("%s:\n", rman->dm_desc);
+
+		len = strlen(rman->dm_desc) + 1;
+		if ((safe_desc = malloc(len)) == NULL) {
+			xo_err(1, "No memory!");
+		}
+		strlcpy(safe_desc, rman->dm_desc, len);
+		xml_safe_string(safe_desc);
+
+		print_indent(indent);
+		xo_emit("{d:%s}:\n", rman->dm_desc);
+		xo_open_list(safe_desc);
 
 		/* print resources */
 		ia->indent = indent + 4;
 		devinfo_foreach_rman_resource(rman,
 		    print_device_matching_resource, ia);
+
+		xo_close_list(safe_desc);
+		free(safe_desc);
 	}
 	ia->indent = indent;
 	return(0);
 }
 
 static void
-print_dev(struct devinfo_dev *dev)
+print_device_props(struct devinfo_dev *dev)
 {
+	if (vflag) {
+		if (*dev->dd_desc) {
+			xo_emit(" <{d:%s}>", dev->dd_desc);
+			xo_emit("{e:description/%s}", dev->dd_desc);
+		}
+		if (*dev->dd_pnpinfo) {
+			xo_open_container("pnpinfo");
+			xo_emit("{D: pnpinfo}");
+
+			if ((strcmp(dev->dd_pnpinfo, "unknown") == 0))
+				xo_emit("{D: unknown}");
+			else
+				print_kvlist(dev->dd_pnpinfo);
+
+			xo_close_container("pnpinfo");
+		}
+		if (*dev->dd_location) {
+			xo_open_container("location");
+			xo_emit("{D: at}");
+			print_kvlist(dev->dd_location);
+			xo_close_container("location");
+		}
+
+		// If verbose, then always print state for json/xml.
+		if (!(dev->dd_flags & DF_ENABLED))
+			xo_emit("{e:state/disabled}");
+		else if (dev->dd_flags & DF_SUSPENDED)
+			xo_emit("{e:state/suspended}");
+		else
+			xo_emit("{e:state/enabled}");
+	}
 
-	printf("%s", dev->dd_name[0] ? dev->dd_name : "unknown");
-	if (vflag && *dev->dd_desc)
-		printf(" <%s>", dev->dd_desc);
-	if (vflag && *dev->dd_pnpinfo)
-		printf(" pnpinfo %s", dev->dd_pnpinfo);
-	if (vflag && *dev->dd_location)
-		printf(" at %s", dev->dd_location);
 	if (!(dev->dd_flags & DF_ENABLED))
-		printf(" (disabled)");
+		xo_emit("{D: (disabled)}");
 	else if (dev->dd_flags & DF_SUSPENDED)
-		printf(" (suspended)");
+		xo_emit("{D: (suspended)}");
 }
 
-
 /*
  * Print information about a device.
  */
-int
+static int
 print_device(struct devinfo_dev *dev, void *arg)
 {
 	struct indent_arg	ia;
-	int			i, indent;
+	int			indent, ret;
+	const char*		devname = dev->dd_name[0] ? dev->dd_name : "unknown";
+	bool			printit = vflag || (dev->dd_name[0] != 0
+					    && dev->dd_state >= DS_ATTACHED);
```

Comment:
we have a preference for && at the end of the prior line rather than the start of this line. And the indent should align the 'dev' under the first 't' in printit.

Reply:
I'm not used to tab-stops indenting. The results seem inconsistent between editors. What editor/config combo can I use to make shure it's right?

Reply:
I have an emacs config file that I've used for a long time. The important part is && at the end. The extra indent is less important, to be honest. If you are using vi/vim/nvi, I'm afraid I can't help you, but maybe tools/tools/editing/freebsd.vim can help? I don't know first hand how well it works.

Reply:
I'm using vscode with no relevant config. But default vi seems more in agreement with github diffs so I'll go with that for now.